### PR TITLE
Add missing Modules to nn.functional

### PR DESCRIPTION
**Pull request by nrbrd**
This PR adds some modules that are not available in nn.functional:

* Dropout2d
* Dropout3d
* BCE
* CosineEmbeddingLoss
* HingeEmbeddingLoss
* L1Loss
* MSELoss
* MarginRankinLoss
* MultilabelMarginLoss
* SoftMarginLoss
* MultilabelSoftMarginLoss
* MultiMarginLoss

### CODE REVIEW

File: test/common_nn.py [hunk 1]
```diff
@@ -217,6 +217,35 @@
 ]
 
 
+functional_criterion_tests = [
```
Comment (fmassa) [inline test/common_nn.py]: I think these functions can go in `test_nn.py`, because I believe `common_nn` is supposed to contain functions that should be run both in `nn` and `legacy_nn`.

File: test/test_nn.py [hunk 2]
```diff
@@ -2606,6 +2607,61 @@ def test_orthogonal(self):
                                          torch.eye(rows) * gain ** 2, prec=1e-6)
 
 
+# functional loss wrappers
+
+
+class _SmoothL1LossFunctional(_Loss):
```
Comment (fmassa) [inline test/test_nn.py]: I'm not sure this is the best way to test them. I actually just noticed that we don't have tests for any of the modules in `functional`, and are relying only on tests for their `Module` counterpart.
I'd say that as the `Module` call into `functional`, in a way that mimics `functional` interface with classes, it would be enough to test `Module` only. @apaszke wdyt?

File: torch/nn/functional.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import torch
 from . import _functions
 from .modules import utils
+from .backends.thnn import backend as thnn_backend
```
Comment (fmassa) [inline torch/nn/functional.py]: @apaszke I find this `backends.thnn` a bit confusing. Is this dead code or half-dead code? I suppose it is there to make room for other backends (like `nnpack` maybe?), but is it still the case? The only use I see (but I'm probably missing something) is that it [renames `FeatureDropout` to `Dropout2d` and `Dropout3d`](https://github.com/pytorch/pytorch/blob/master/torch/nn/backends/thnn.py#L40-L41).

**Review by fmassa — COMMENTED** [review]
This overall looks very good, thanks Aron!
The main remark I have is about the tests, but I'll let @apaszke comment on that.

File: test/test_nn.py (hunk 2 above)
Comment (hughperkins) [inline test/test_nn.py]: Depends on the implementations, I think? Kind of funny though if we have a situation hypothetically where one implementation wraps the other, and both tests simply compare the Module result with the functional result :-)

**Comment by szagoruyko** [comment]
@aron-bordin any progress? should I take over and finish it?

**Comment by aron-bordin** [comment]
I'll rebase and update it. I was expecting the answer about how I implemented the tests, using a fake module, but I'll move on.

File: test/test_nn.py (hunk 2 above)
Comment (aron-bordin) [inline test/test_nn.py]: I added this interface to test these functional functions because they are not called in their corresponding `Module`. 

The Module uses the class name to get the function directly from the backend (as https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py#L22 and https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py#L33)

I don't know what is the best way to do it: (1) using this "fake modules" to test, (2) update the nn.loss to actually call the functional instead of using the class name to get the function from the backend (as currently implemented) or (3) something else.

Comment (szagoruyko) [inline test/test_nn.py]: def (2), that's what we do for all other nn modules

**Comment by aron-bordin** [comment]
Hi, I updated the PR, it's ready to be reviewed again ;)
Thx.
cc: @fmassa @hughperkins

**Comment by szagoruyko** [comment]
looks good

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks a lot Aron!
